packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,164 @@
+// with perfomance entry name. Additionally, the entry's start/end
+// timings must fit with in the span's start/end timings.
+export function getPerfResourceEntries(
+  xhrUrl: string,
+  span: Span
+): PerformanceResourceTiming[] {
+  return performance
+    .getEntriesByType('resource')
+    .filter(entry =>
+      isPerfEntryPartOfXhr(entry as PerformanceResourceTiming, xhrUrl, span)
+    ) as PerformanceResourceTiming[];
+}
+
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let perfEntry1: PerformanceResourceTiming;
+  let perfEntry2: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
+  // this array is not big as the performance resource entries is cleared
+  // when there are no more running XHRs.
+  for (let i = 0; i < perfEntries.length; i++) {
+    perfEntry1 = perfEntries[i];
+    // Compare every performance entry with its consecutive perfomance entries.
+    // That way to avoid comparing twice the entries.
+    for (let j = i + 1; j < perfEntries.length; j++) {
+      perfEntry2 = perfEntries[j];
+      if (!overlappingPerfResourceTimings(perfEntry1, perfEntry2)) {
+        // As the entries are not overlapping, that means those timings
+        // are possible perfomance timings related to the XHR.
+        possiblePerfEntries.push([perfEntry1, perfEntry2]);
+        pairedEntries.add(perfEntry1);
+        pairedEntries.add(perfEntry2);
+      }
+    }
+    // If the entry1 couldn't be paired with any other resource timing,
+    // add it as a single resource timing. This is possible because this
+    // single entry might be better that the other possible entries.
+    if (!pairedEntries.has(perfEntry1)) {
+      possiblePerfEntries.push(perfEntry1 as PerformanceResourceTiming);
+    }
+  }
+  return possiblePerfEntries;
+}
+
+// The best Performance Resource Timing Entry is considered the one with the
+// minimum gap the span end/start timings. That way we think that it fits
+// better to the XHR as it is the closest data to the actual XHR.
+function getBestPerfResourceTiming(
+  perfEntries: XhrPerformanceResourceTiming[],
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  let minimumGapToSpan = Number.MAX_VALUE;
+  let bestPerfEntry: XhrPerformanceResourceTiming | undefined = undefined;
+  let sumGapsToSpan: number;
+  for (const perfEntry of perfEntries) {
+    // As a Tuple is in the end an Array, check that perfEntry is instance of
+    // Array is enough to know if this value refers to a Tuple.
+    if (perfEntry instanceof Array) {
+      sumGapsToSpan = Math.abs(perfEntry[0].startTime - span.startPerfTime);
+      sumGapsToSpan += Math.abs(perfEntry[1].responseEnd - span.endPerfTime);
```

Comment:
Could you do something like `perfEntry[perfEntry.length - 1]` here and remove the `} else {` block by making it so that the paired case (with CORS preflight) and non-pair cases both use a list?